### PR TITLE
Add progress bar

### PR DESCRIPTION
**Pull request by abocquet**
Adds a progress bar to mesolve and sesolve. Performances are not optimal yet since an event is triggered at each timestep

### CODE REVIEW

File: dynamiqs/mesolve/lindblad_term.py [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
 
-class LindbladTerm(dx.ODETerm):
+class LindbladTerm(ProgressBarTerm):
     H: TimeArray  # (n, n)
     Ls: list[TimeArray]  # (nL, n, n)
     vector_field: Callable[[Scalar, PyTree, PyTree], PyTree]
 
-    def __init__(self, H: TimeArray, Ls: TimeArray):
+    def __init__(
+        self, H: TimeArray, Ls: TimeArray, update_progressbar: Optional[Callable]
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/lindblad_term.py:18]: New-gen typing
```suggestion
        self, H: TimeArray, Ls: TimeArray, update_progressbar: callable | None
```

File: dynamiqs/mesolve/mesolve.py [hunk 2]
```diff
@@ -42,11 +43,14 @@ def mesolve(
     # === stepsize controller
     stepsize_controller, dt = _stepsize_controller(solver)
 
+    # === progress bar
+    bar, update_progressbar = make_progressbar(options.progress_bar, tsave[-1])
```
Comment (pierreguilmin) [inline dynamiqs/mesolve/mesolve.py:47]: We maybe need to specify `tsave[0]` too bc we solve from `t0=tsave[0]`? Ideally `t0` should be an option in `options`.

File: dynamiqs/options.py [hunk 3]
```diff
@@ -44,6 +44,7 @@ def __init__(
         verbose: bool = True,
         dtype: jnp.complex64 | jnp.complex128 | None = None,
         cartesian_batching: bool = True,
+        progress_bar: bool = True,
```
Comment (pierreguilmin) [inline dynamiqs/options.py:47]: This is redundant with `verbose`. `progress_bar` is more explicit but longer, pick what you prefer!

File: dynamiqs/progress_bar.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+from __future__ import annotations
+
+from typing import Callable, Union
```
Comment (pierreguilmin) [inline dynamiqs/progress_bar.py:3]: New-gen typing, we probably don't need this

Callable -> callable
Union[a, b] -> a | b

Comment (pierreguilmin) [reply dynamiqs/progress_bar.py:3]: This is why we `from __future__ import annotations`

File: dynamiqs/progress_bar.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+    def __init__(self, update_progressbar: Union[Callable, None]):
+        self.update_progressbar = update_progressbar
+
+    def vector_field(self, t: Scalar, _psi: PyTree, _args: PyTree):
+        if self.update_progressbar is not None:
+            host_callback.id_tap(self.update_progressbar, t)
+
+
+def make_progressbar(progress_bar: bool, tfinal: float):
```
Comment (pierreguilmin) [inline dynamiqs/progress_bar.py:22]: Can yout type the return?

**Review by pierreguilmin — DISMISSED** [review]
Nice!  Ideally we push this upstream to diffrax directly.

File: dynamiqs/progress_bar.py (hunk 4 above)
Comment (abocquet) [reply dynamiqs/progress_bar.py:3]: TIL

File: dynamiqs/progress_bar.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
+from __future__ import annotations
+
+
+import diffrax as dx
+from jaxtyping import PyTree, Scalar
+from jax.experimental import host_callback
+from tqdm import tqdm
```
Comment (gautierronan) [inline dynamiqs/progress_bar.py:7]:
```suggestion
from tqdm.auto import tqdm
```
(to handle both notebooks and terminals)

**Pull request closed without merging**